File: client-go/defaults.go
package queen

// Default configuration values for Queen Client.
// Convention:
// - Properties with "Millis" suffix → milliseconds
// - Properties with "Seconds" suffix → seconds
// - Properties without suffix (time-related) → seconds

// ClientDefaults contains default values for client configuration.
var ClientDefaults = ClientConfig{
	TimeoutMillis:           30000,    // 30 seconds
	RetryAttempts:           3,        // 3 retry attempts
	RetryDelayMillis:        1000,     // 1 second initial delay (exponential backoff)
	LoadBalancingStrategy:   "affinity", // 'round-robin', 'session', or 'affinity'
	AffinityHashRing:        128,      // Number of virtual nodes per server for affinity strategy
	EnableFailover:          true,     // Auto-failover to other servers
	HealthRetryAfterMillis:  5000,     // Retry unhealthy backends after 5 seconds
}

// QueueDefaults contains default values for queue configuration.
var QueueDefaults = QueueConfig{
	LeaseTime:                 300,   // 5 minutes (seconds)
	RetryLimit:                3,     // Max 3 retries before DLQ
	Priority:                  0,     // Default priority
	DelayedProcessing:         0,     // No delay (seconds)
	WindowBuffer:              0,     // No window buffering (seconds)
	MaxSize:                   0,     // No limit on messages per queue
	RetentionSeconds:          0,     // No retention (keep forever)
	CompletedRetentionSeconds: 0,     // No retention for completed messages
	EncryptionEnabled:         false, // No encryption by default
}

// ConsumeDefaults contains default values for consume operations.
var ConsumeDefaults = ConsumeOptions{
	Concurrency:   1,     // Single worker
	Batch:         1,     // One message at a time
	AutoAck:       true,  // Client-side auto-ack (NOT sent to server)
	Wait:          true,  // Long polling enabled
	TimeoutMillis: 30000, // 30 seconds long poll timeout
	Limit:         0,     // No limit (run forever) - 0 means no limit
	IdleMillis:    0,     // No idle timeout - 0 means no timeout
	RenewLease:    false, // No auto-renewal
	Each:          false, // Process as batch by default
}

// PopDefaults contains default values for pop operations.
var PopDefaults = PopOptions{
	Batch:         1,     // One message
	Wait:          false, // No long polling (immediate return)
	TimeoutMillis: 30000, // 30 seconds if wait=true
	AutoAck:       false, // Server-side auto-ack (false = manual ack required)
}

// BufferDefaults contains default values for buffer configuration.
var BufferDefaults = BufferConfig{
	MessageCount: 100,  // Flush after 100 messages
	TimeMillis:   1000, // Or flush after 1 second
}

// DefaultPartition is the default partition name.
const DefaultPartition = "Default"

// QueueModeConsumerGroup is the consumer group name used when no group is specified.
const QueueModeConsumerGroup = "__QUEUE_MODE__"

// AckStatusCompleted is the status for successful message processing.
const AckStatusCompleted = "completed"

// AckStatusFailed is the status for failed message processing.
const AckStatusFailed = "failed"

// SubscriptionModeAll includes all messages (historical + new).
const SubscriptionModeAll = "all"

// SubscriptionModeNew skips historical messages, starts from new.
const SubscriptionModeNew = "new"

// SubscriptionModeNewOnly is an alias for SubscriptionModeNew.
const SubscriptionModeNewOnly = "new-only"

// SubscriptionFromNow starts subscription from current time.
const SubscriptionFromNow = "now"

File: client-js/src/utils/defaults.ts
/**
 * Default configuration values for Queen Client
 * Following the convention:
 * - Properties with "Millis" suffix -> milliseconds
 * - Properties with "Seconds" suffix -> seconds
 * - Properties without suffix (time-related) -> seconds
 */

export const CLIENT_DEFAULTS = {
  timeoutMillis: 30000,
  retryAttempts: 3,
  retryDelayMillis: 1000,
  loadBalancingStrategy: 'affinity' as const,
  affinityHashRing: 128,
  enableFailover: true,
  healthRetryAfterMillis: 5000,
  bearerToken: null as string | null,
  headers: {} as Record<string, string>
}

export const QUEUE_DEFAULTS = {
  leaseTime: 300,
  retryLimit: 3,
  priority: 0,
  delayedProcessing: 0,
  windowBuffer: 0,
  maxSize: 0,
  retentionSeconds: 0,
  completedRetentionSeconds: 0,
  encryptionEnabled: false
}

export const CONSUME_DEFAULTS = {
  concurrency: 1,
  batch: 1,
  autoAck: true,
  wait: true,
  timeoutMillis: 30000,
  limit: null as number | null,
  idleMillis: null as number | null,
  renewLease: false,
  renewLeaseIntervalMillis: null as number | null,
  subscriptionMode: null as string | null,
  subscriptionFrom: null as string | null
}

export const POP_DEFAULTS = {
  batch: 1,
  wait: false,
  timeoutMillis: 30000,
  autoAck: false
}

export const BUFFER_DEFAULTS = {
  messageCount: 100,
  timeMillis: 1000
}
